### DISCUSSION TITLE
next-translate

### DISCUSSION THREAD

**Discussion by rsipakov** [timlrx/tailwind-nextjs-starter-blog#92] [Ideas]
Hey there! I am trying to integrate internalization for this project using the "[next-translate](https://github.com/vinissimus/next-translate#readme)" plugin. I understand how to translate basic pages, but I don't know how to deal with blog pages wich are use .md and .mdx formats. Can somebody help to implement this future for the blog pages?

<a href="https://www.loom.com/share/56f6039384b940079cde34f674bb84a8"> <p>Next.js Starter Blog + next-translate - Watch Video</p> <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/56f6039384b940079cde34f674bb84a8-with-play.gif"> </a>

**Comment by timlrx** [comment]
Can't really help since I have no experience with the package but if I were to try something, I guess the translate function would have to act on [mdxSource returned at the blog page](https://github.com/timlrx/tailwind-nextjs-starter-blog/blob/master/pages/blog/%5B...slug%5D.js#L36)

**Reply by timlrx** [reply]
This [stackoverflow](https://stackoverflow.com/questions/53119525/how-to-use-i18next-react-i18next-inside-mdx-markdown-files-with-mdxjs) post should be helpful but seem quite tedious. An alternative approach would be to have different translated markdown files in separate folders within `data/blog` e.g. `en` or `ru` and detect the users language and display the correct post. This is supported with our nested route approach.

**Reply by rsipakov** [reply]
Hi! Thank you for your advice. I will try with separate folders.

**Comment by GautierArcin** [comment]
Is it still under discussion ? I might be interested to also participate, I might need it for my personnal blog.

**Reply by timlrx** [reply]
Hi, no plans to include i18n into the template, but it should be reasonably easy to add `next-translate` and separate the markdown content in different language folders or change the parsing of the markdown slug to detect different file extensions like `blog-post.en.md` vs `blog.post.ru.md`

If anyone has it nicely setup, I will be happy to link it as an example in the readme.

**Reply by vahidtakro** [reply]
> Is it still under discussion ? I might be interested to also participate, I might need it for my personnal blog.

I am also busy with this,
what I could do so far was the changes I've done to mdx.js file, nothing changed
```
 const {defaultlocale} = require('../i18n.json')
.
.
.
export async function getFileBySlug(type, slug, locale) {
 const filename = locale === defaultlocale ? 'index.md' : `index.${locale}.md`
 const filenameMdx = locale === defaultlocale ? 'index.mdx' : `index.${locale}.mdx`

  const mdxPath = path.join(root, 'data', type, `${slug}`, filenameMdx)
  const mdPath = path.join(root, 'data', type, `${slug}`, filename)
  const source = fs.existsSync(mdxPath)
    ? fs.readFileSync(mdxPath, 'utf8')
    : fs.readFileSync(mdPath, 'utf8')
```

if somebody knows what changes, and where should I change, to internationalize my markdown blog, I will be thankful, I am also using next-translate for basic titles, pages, etc. but still for markdown part, some changes need to be done in the codes.

**Comment by timlrx** [comment]
Take a look at the demo/translate branch for an example of getting i18n to work and translating the blog posts. Here's the deployed website:

- [English](https://tailwind-nextjs-starter-blog-5s30wfqd4-timlrx.vercel.app/)
- [French](https://tailwind-nextjs-starter-blog-5s30wfqd4-timlrx.vercel.app/fr)

More information in #160. Hope that helps!

**Reply by rsipakov** [reply]
Did you @timlrx do it without additional packages such as example "next-translate", "react-i18next" or "react-intl"? Can you advise how can I do translate for title and menu navigation in this case? Thank you!

**Reply by rsipakov** [reply]
I tried adding the package "next-translate". It seems to be working. Can you @timlrx create a new branch demo/next-translate for this purpose? [demo/next-translate](https://github.com/rsipakov/tailwind-nextjs-starter-blog/tree/demo/next-translate)
p.s. Also, I need a help to modify next.config.js to implement [next-compose-plugins](https://github.com/cyrilwanner/next-compose-plugins)

**Reply by rsipakov** [reply]
p.s. demo https://tailwind-nextjs-starter-blog-rsipakov-23c6jgjfk-jamstack-os.vercel.app/

**Reply by rsipakov** [reply]
... with "next-compose-plugins"

```
const withBundleAnalyzer = require('@next/bundle-analyzer')({
  enabled: process.env.ANALYZE === 'true',
})
const nextTranslate = require('next-translate')
const withPlugins = require('next-compose-plugins')

module.exports = withPlugins([[withBundleAnalyzer], [nextTranslate]], {
  reactStrictMode: true,
  pageExtensions: ['js', 'jsx', 'md', 'mdx'],
  eslint: {
    dirs: ['pages', 'components', 'lib', 'layouts', 'scripts'],
  },
  webpack: (config, { dev, isServer }) => {
    config.module.rules.push({
      test: /\.(png|jpe?g|gif|mp4)$/i,
      use: [
        {
          loader: 'file-loader',
          options: {
            publicPath: '/_next',
            name: 'static/media/[name].[hash].[ext]',
          },
        },
      ],
    })

    config.module.rules.push({
      test: /\.svg$/,
      use: ['@svgr/webpack'],
    })

    if (!dev && !isServer) {
      // Replace React with Preact only in client production build
      Object.assign(config.resolve.alias, {
        react: 'preact/compat',
        'react-dom/test-utils': 'preact/test-utils',
        'react-dom': 'preact/compat',
      })
    }

    return config
  },
})
```

or without "next-compose-plugins"

```
module.exports = nextTranslate(
  withBundleAnalyzer({
    reactStrictMode: true,
....
```

@timlrx What is a better way?

**Reply by timlrx** [reply]
Nice job adding on to my demo! Yes, I just used [nextjs default i18n routing](https://nextjs.org/docs/advanced-features/i18n-routing) with no additional packages. But, to manage the different content for the title, menu navigation etc., it might be easier to use one of those packages you listed.

No need for a new demo branch? I don't mind merging your demo with next-translate to my demo/translate branch. 

Using `next-compose-plugins` is probably cleaner.

**Reply by GautierArcin** [reply]
Nice ! I think `next-translate` is indeed a good (and lightweight) choice. I tried to do the internalization of the blog with `next-i18next` before, and encountered some issues because of webpack version ([see here](https://github.com/isaachinman/next-i18next/issues/879)).

@rsipakov I added navBar translation and translation of the `About` page on your branch. [Live preview is here](https://tailwind-nextjs-starter-blog-rsipakov-git-fo-6a9bc9-jamstack-os.vercel.app/). I'm not sure about the `About` page though: it might be smarter to do it with SSR now. What do you think ?

I think I might begin to do the full internalization (with `en` and `fr` locales) of my blog pretty soon Thanks a lot for the help both of you provided. 😃

**Reply by vahidtakro** [reply]
thanks! nice job! I think there would be some bugs on it,
tags will lead to 404 error, also second, third, ... pages of the blog will lead to 404.
also if you want to post only in one language, the post in another language will lead to error, not 404. Am I right?
in this case there need to be a lot of changes, also for author pages, sitemap, seo, rss generators.

**Reply by timlrx** [reply]
@GautierArcin happy to add a link in the readme to your repo when you are done. Thanks to you and @rsipakov as well for chipping in. Not sure, whether we should enable more "out of the box" support for i18n since it makes the codebase more opinionated and less flexible, but having a model reference will surely help people who are looking to achieve the same!

@vahidtakro you have to make similar adjustments to the tags page for it to work. I updated my demo/translate to patch that. Yes, this setup assumes that you have a translated blog post for each post you create. Yes, i18n is relatively tedious, since you have to translate each piece of content / text that exist.

**Reply by vahidtakro** [reply]
> @GautierArcin happy to add a link in the readme to your repo when you are done. Thanks to you and @rsipakov as well for chipping in. Not sure, whether we should enable more "out of the box" support for i18n since it makes the codebase more opinionated and less flexible, but having a model reference will surely help people who are looking to achieve the same!
> 
> @vahidtakro you have to make similar adjustments to the tags page for it to work. I updated my demo/translate to patch that. Yes, this setup assumes that you have a translated blog post for each post you create. Yes, i18n is relatively tedious, since you have to translate each piece of content / text that exist.

Thanks! I see no more problem with opening tags, the problem is only tag counts, and "tag" page shows all tags in any other languages, and the number of tags are wrong, so if I write one post in two languages with same tag, it shows --tag (2)--
(for now all tags must be similar to work, I mean common tags for both languages, which is not a good logic)

also the biggest problem on this demo is second page of blog, which should be related to pages/blog/page/[page].js I think, it gets 404 when going to next pages, and author pages are not translated, (for me I could do it, with the code below, but still when I go to translated posts where authordetails are, it's not showing the translated author)
and also, when a post is only available in one language, there will be a problem, maybe there can be a md 404 default page to refer, with the posts that are not available in other language?
I think this is easy to fix, but I am not so professional to fix. maybe one of you guys can? @rsipakov @timlrx @GautierArcin 

thanks.

about.js file changes:
```
export async function getStaticProps({ locale, filename }) {
  filename = locale === 'en' ? ['default'] : ['default.fr']
  const authorDetails = await getFileBySlug('authors', filename)
  return { props: { authorDetails } }
}
```
also @GautierArcin changed it like this:
```
export async function getStaticProps({ locale, defaultLocale }) {
  const authorDetails = await getFileBySlug('authors', [
    `default${defaultLocale !== locale ? `.${locale}` : ''}`,
  ])
  return { props: { authorDetails } }
}
```

**Reply by rsipakov** [reply]
> Nice job adding on to my demo! Yes, I just used [nextjs default i18n routing](https://nextjs.org/docs/advanced-features/i18n-routing) with no additional packages. But, to manage the different content for the title, menu navigation etc., it might be easier to use one of those packages you listed.
> 
> No need for a new demo branch? I don't mind merging your demo with next-translate to my demo/translate branch.
> 
> Using `next-compose-plugins` is probably cleaner.

Thank you for your feedback!)

**Reply by rsipakov** [reply]
> Nice ! I think `next-translate` is indeed a good (and lightweight) choice. I tried to do the internalization of the blog with `next-i18next` before, and encountered some issues because of webpack version ([see here](https://github.com/isaachinman/next-i18next/issues/879)).
> 
> @rsipakov I added navBar translation and translation of the `About` page on your branch. [Live preview is here](https://tailwind-nextjs-starter-blog-rsipakov-git-fo-6a9bc9-jamstack-os.vercel.app/). I'm not sure about the `About` page though: it might be smarter to do it with SSR now. What do you think ?
> 
> I think I might begin to do the full internalization (with `en` and `fr` locales) of my blog pretty soon Thanks a lot for the help both of you provided. 😃

@GautierArcin Thank you for your feedback! ;)  I tried to do with "react-intl" before too, and I believe that "next-translate" better. Also, in this case you can use for manage .json this service (Translation editor) [BabelEdit](https://www.codeandweb.com/babeledit)

**Reply by rsipakov** [reply]
> Nice ! I think `next-translate` is indeed a good (and lightweight) choice. I tried to do the internalization of the blog with `next-i18next` before, and encountered some issues because of webpack version ([see here](https://github.com/isaachinman/next-i18next/issues/879)).
> 
> @rsipakov I added navBar translation and translation of the `About` page on your branch. [Live preview is here](https://tailwind-nextjs-starter-blog-rsipakov-git-fo-6a9bc9-jamstack-os.vercel.app/). I'm not sure about the `About` page though: it might be smarter to do it with SSR now. What do you think ?
> 
> I think I might begin to do the full internalization (with `en` and `fr` locales) of my blog pretty soon Thanks a lot for the help both of you provided. 😃

@GautierArcin "I'm not sure about the About page though: it might be smarter to do it with SSR now. What do you think ?" - for this question I will back on Sunday.

**Comment by GautierArcin** [comment]
Hello everyone,

I finished a first version of the translation. Locales are `en` and `fr`. [Sources branch github](https://github.com/GautierArcin/tailwind-nextjs-starter-blog/tree/demo/next-translate) & [Preview ]( https://tailwind-nextjs-starter-blog-seven.vercel.app/ ). 

I'm not super sure about some part of my code (using imperative code instead of functional in some `getStaticProps()`, for example, or maybe using SSG instead of hooks, for performances). I also tried to made it as generic as possible, replacing hard-coded `en-us` by `defaultLangage` when possible.

There two kinds of behaviors:
1. You can have a fully "multi-langage" SSR blog. If there is an article `article.mdx`, you can put an article for each language defined in languages (for example, it would be `article.fr.mdx` for french and `articles.mdx` for english if there is [`fr`,`en`] locales & defaultLocale is `en` ).
2. When there is no corresponding article for the locales ( in our previous case, `article.fr.mdx` is missing), the blog is gonna use the defaultLocale article for the article corresponding to the missing locale.  To be more explicit, in my case, I plan to write my blog in english, therefore I plan to use `en ` as default language.

@timlrx  

I think it's a good idea to just keep the branch as a demo, because my implementation is pretty opinionated.
At the same time, the second behavior might allow to just put one locale (`english`) and use it as defaultLocale. Therefore, for user that wants only an english blog, it would not change anything, but users that wants a blog in their langage would be able to do so in minimum time (2 to 3 hours ?)

 Let me know what you think ?

(Totally unrelated, but you begun an implementation of analytics ? I'm interested in it 😅)

@vahidtakro  @rsipakov It should be better, let me know what you think.

Let me know what you think  / what I might have missed.

**Reply by vahidtakro** [reply]
Hi @GautierArcin I will just try it now :)

**Reply by vahidtakro** [reply]
well done!
I now comment on this,

I think the logic "When there is no corresponding article for the locales ( in our previous case, article.fr.mdx is missing), the blog is gonna use the defaultLocale article for the article corresponding to the missing locale." is not a good idea due to SEO, I think it's better to don't display that post in the "FR" language at all, it shouldn't be available, in any case. if you check how "Hugo CMS" work, this is the best logic, when you switch to other language, the content should be different, it's like two different websites. and this is what everyone likes.

I am still checking and looking for problems! but anyway it's a good job! bravo!

What do you mean about " implementation of analytics " you mean real-time post views? , I also think of a self-hosted comment system, without github login.

**Reply by vahidtakro** [reply]
how did you manage to show tags only related to the locale posts in "tag" page? which files did you change for this?

**Reply by rsipakov** [reply]
> Hello everyone,
> 
> I finished a first version of the translation. Locales are `en` and `fr`. [Sources branch github](https://github.com/GautierArcin/tailwind-nextjs-starter-blog/tree/demo/next-translate) & [Preview ](https://tailwind-nextjs-starter-blog-seven.vercel.app/).
> 
> I'm not super sure about some part of my code (using imperative code instead of functional in some `getStaticProps()`, for example, or maybe using SSG instead of hooks, for performances). I also tried to made it as generic as possible, replacing hard-coded `en-us` by `defaultLangage` when possible.
> 
> There two kinds of behaviors:
> 
> 1. You can have a fully "multi-langage" SSR blog. If there is an article `article.mdx`, you can put an article for each language defined in languages (for example, it would be `article.fr.mdx` for french and `articles.mdx` for english if there is [`fr`,`en`] locales & defaultLocale is `en` ).
> 2. When there is no corresponding article for the locales ( in our previous case, `article.fr.mdx` is missing), the blog is gonna use the defaultLocale article for the article corresponding to the missing locale.  To be more explicit, in my case, I plan to write my blog in english, therefore I plan to use `en ` as default language.
> 
> @timlrx
> 
> I think it's a good idea to just keep the branch as a demo, because my implementation is pretty opinionated.
> At the same time, the second behavior might allow to just put one locale (`english`) and use it as defaultLocale. Therefore, for user that wants only an english blog, it would not change anything, but users that wants a blog in their langage would be able to do so in minimum time (2 to 3 hours ?)
> 
> Let me know what you think ?
> 
> (Totally unrelated, but you begun an implementation of analytics ? I'm interested in it 😅)
> 
> @vahidtakro @rsipakov It should be better, let me know what you think.
> 
> Let me know what you think / what I might have missed.

@timlrx @GautierArcin I agree with this "...good idea to just keep the branch as a demo".

**Reply by rsipakov** [reply]
> Hello everyone,
> 
> I finished a first version of the translation. Locales are `en` and `fr`. [Sources branch github](https://github.com/GautierArcin/tailwind-nextjs-starter-blog/tree/demo/next-translate) & [Preview ](https://tailwind-nextjs-starter-blog-seven.vercel.app/).
> 
> I'm not super sure about some part of my code (using imperative code instead of functional in some `getStaticProps()`, for example, or maybe using SSG instead of hooks, for performances). I also tried to made it as generic as possible, replacing hard-coded `en-us` by `defaultLangage` when possible.
> 
> There two kinds of behaviors:
> 
> 1. You can have a fully "multi-langage" SSR blog. If there is an article `article.mdx`, you can put an article for each language defined in languages (for example, it would be `article.fr.mdx` for french and `articles.mdx` for english if there is [`fr`,`en`] locales & defaultLocale is `en` ).
> 2. When there is no corresponding article for the locales ( in our previous case, `article.fr.mdx` is missing), the blog is gonna use the defaultLocale article for the article corresponding to the missing locale.  To be more explicit, in my case, I plan to write my blog in english, therefore I plan to use `en ` as default language.
> 
> @timlrx
> 
> I think it's a good idea to just keep the branch as a demo, because my implementation is pretty opinionated.
> At the same time, the second behavior might allow to just put one locale (`english`) and use it as defaultLocale. Therefore, for user that wants only an english blog, it would not change anything, but users that wants a blog in their langage would be able to do so in minimum time (2 to 3 hours ?)
> 
> Let me know what you think ?
> 
> (Totally unrelated, but you begun an implementation of analytics ? I'm interested in it 😅)
> 
> @vahidtakro @rsipakov It should be better, let me know what you think.
> 
> Let me know what you think / what I might have missed.

@timlrx @GautierArcin I canceled my pull request for demo/next-translate. And I think is better way when you @GautierArcin make your pull request to demo/translate as mentioned before @timlrx.

**Reply by rsipakov** [reply]
> well done!
> I now comment on this,
> 
> I think the logic "When there is no corresponding article for the locales ( in our previous case, article.fr.mdx is missing), the blog is gonna use the defaultLocale article for the article corresponding to the missing locale." is not a good idea due to SEO, I think it's better to don't display that post in the "FR" language at all, it shouldn't be available, in any case. if you check how "Hugo CMS" work, this is the best logic, when you switch to other language, the content should be different, it's like two different websites. and this is what everyone likes.
> 
> I am still checking and looking for problems! but anyway it's a good job! bravo!
> 
> What do you mean about " implementation of analytics " you mean real-time post views? , I also think of a self-hosted comment system, without github login.

@vahidtakro - I'm agree with this explanation for SEO. Maybe is good idea if user will indicate notice for not translated content like ("in progress for translate").

**Reply by vahidtakro** [reply]
> > well done!
> > I now comment on this,
> > 
> > I think the logic "When there is no corresponding article for the locales ( in our previous case, article.fr.mdx is missing), the blog is gonna use the defaultLocale article for the article corresponding to the missing locale." is not a good idea due to SEO, I think it's better to don't display that post in the "FR" language at all, it shouldn't be available, in any case. if you check how "Hugo CMS" work, this is the best logic, when you switch to other language, the content should be different, it's like two different websites. and this is what everyone likes.
> > 
> > I am still checking and looking for problems! but anyway it's a good job! bravo!
> > 
> > What do you mean about " implementation of analytics " you mean real-time post views? , I also think of a self-hosted comment system, without github login.
> 
> @vahidtakro - I'm agree with this explanation for SEO. Maybe is good idea if user will indicate notice for not translated content like ("in progress for translate").

I think being unavailable is the best, nothing, maybe a 404 error. Just like how multilingual Hugo works. Everything must be separated in both languages. It only becomes related if posts are common in names with (locale).md at the end.
So one can write single posts in different languages too.
Different categories, different tags.
I mean no words in the locale should be seen on the other.

The way if we add a file for "no translation is available" i think search engines will list this page as well.
It just shouldn't go on sitemap, if the post is not available in that language.

**Reply by timlrx** [reply]
@GautierArcin nice work getting everything translated :) I agree with keeping it as a demo. Will just link to your repo from the readme. 

Agree with both of them that if a post for a particular language is missing, there's no need to display it (i.e. no fallback seems like a more natural behaviour). 

As for the sitemap, [this stackexchange post](https://webmasters.stackexchange.com/questions/125236/sitemap-best-practice-for-multilingual-website-should-i-list-my-urls-in-each-l) gives good guidance on the correct format. It seems that there should be one sitemap with all the URLs. For URLs with multiple translations, there is an additional `xhtml:link` tag. So a particular URL would look like this:

```
<url>
    <loc>http://www.example.com/english/page1.html</loc>
    <xhtml:link 
               rel="alternate"
               hreflang="de"
               href="http://www.example.com/deutsch/page1.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="de-ch"
               href="http://www.example.com/schweiz-deutsch/page1.html"/>
    <xhtml:link 
               rel="alternate"
               hreflang="en"
               href="http://www.example.com/english/page1.html"/>
  </url>
  ```

**Reply by timlrx** [reply]
analytics has already been implemented! You just need to update `siteMetadata` as shown in [my site's config](https://github.com/timlrx/timlrx.com/blob/master/data/siteMetadata.js#L16) to use any of the providers. Give it a try or open another discussion on this - this thread is long enough😅

**Reply by vahidtakro** [reply]
@GautierArcin when I clone this repo, or I integrate it with mine, I don't know why posts are lost, "it shows: No posts found." why is it like that?
it works on github, but not on localhost, on localhosts posts are not rendered, and this is only about this demo, when I clone @timlrx repo, this is not happening, everything works well.
![Captlure](https://user-images.githubusercontent.com/40500566/128605179-241532f9-c299-4ff3-bba2-2a4b4b575fba.JPG)

**Reply by GautierArcin** [reply]
Thank you all !

@vahidtakro I just cloned my repo to test it again and can't replicate the bug. It works as intended on my end. Did you clone my repo and checkout to the right branch ( `demo/next-translate` ) ? I'm using `yarn` if that helps.

As for the `Tags`, look into the `getStaticProps` function of the `Tags` page.

```javascript

export async function getStaticProps({ defaultLocale, locale }) {
  const otherLocale = locale !== defaultLocale ? locale : ''
  const tags = await getAllTags('blog', otherLocale)

  return { props: { tags } }
}

```

@timlrx @rsipakov  Understood for the fallback. I'll do the change next week, it should be easy (It's just three function to edit in `mdx.js` I think). I'll do the PR after everything is in place. 😄

Thanks for the direction for the sitemap ! Shall I modify `scripts/generate-sitemap.js` ? I might need your help if I don't suceed by myself. 😅

Also, I'm thinking to implement is using [`deepl` API](https://www.deepl.com/docs-api/) to automatically translate to other locales. I'll try it during next few months, I don't know how hard it will be and even if it's possible. 😅 😅 

( @timlrx  Thank you a lot for your answer concerning analytics ! @rsipakov Totally unrelated, but how did you do the newsletter on your blog ? I'm sorry to pollute the discussion to such extents, there's no way to direct message people on github. 😅 )

**Reply by vahidtakro** [reply]
> Thank you all !
> 
> @vahidtakro I just cloned my repo to test it again and can't replicate the bug. It works as intended on my end. Did you clone my repo and checkout to the right branch ( `demo/next-translate` ) ? I'm using `yarn` if that helps.
> 
> As for the `Tags`, look into the `getStaticProps` function of the `Tags` page.
> 
> ```js
> export async function getStaticProps({ defaultLocale, locale }) {
>   const otherLocale = locale !== defaultLocale ? locale : ''
>   const tags = await getAllTags('blog', otherLocale)
> 
>   return { props: { tags } }
> }
> ```
> 
> @timlrx @rsipakov Understood for the fallback. I'll do the change next week, it should be easy (It's just three function to edit in `mdx.js` I think). I'll do the PR after everything is in place. 😄
> 
> Thanks for the direction for the sitemap ! Shall I modify `scripts/generate-sitemap.js` ? I might need your help if I don't suceed by myself. 😅
> 
> Also, I'm thinking to implement is using [`deepl` API](https://www.deepl.com/docs-api/) to automatically translate to other locales. I'll try it during next few months, I don't know how hard it will be and even if it's possible. 😅 😅
> 
> ( @timlrx Thank you a lot for your answer concerning analytics ! @rsipakov Totally unrelated, but how did you do the newsletter on your blog ? I'm sorry to pollute the discussion to such extents, there's no way to direct message people on github. 😅 )

when I clone it as fresh as it is now, it works fine, I checked everything for hours now. I thought there must be a problem in my codes, but once I add more posts and tags, this happens. but I found the problem right now, when I remove this code in mdx.js and replace it with the original one, it all works again:

```
export async function getAllFilesFrontMatter(folder, otherLocale) {
  const prefixPaths = path.join(root, 'data', folder)

  const allFiles = getAllFilesRecursively(prefixPaths)
  let files = allFiles.filter((path) => (path.match(/\./g) || []).length === 1)

  // Check if the file exist in the otherlocale. If not, fallback to defaultLangage
  if (otherLocale !== '') {
    files = files.map((e) =>
      allFiles.indexOf(e.replace('.md', `.${otherLocale}.md`)) > -1
        ? e.replace('.md', `.${otherLocale}.md`)
        : e
    )
  }
```
but it shows all the posts. in any language.
there must be a problem with this code I guess.

**Reply by GautierArcin** [reply]
@vahidtakro  Can you upload a version with the bug on git and link it here ?  I tried again to replicate the bug for 30 minutes to no avail. Otherwise, I'm sorry, but I'm afraid I won't be able to help you. 

The code you showed (that is no more present, I changed the behavior) just says that the static generation tries to get every article ending in `.md`(so not the article ending in `.locale.md`).
If the article is only present in `.locale.md`, the blog wasn't able to detect it. That's old behavior and I changed it two hours ago though.

**Reply by vahidtakro** [reply]
Thanks, I saw the changes, still is not working on localhost, if you just download the repo. as the repo is private, I just gave you the access. It's very strange, I'm busy with it for more than 24 hours already. can't find the reason, and this happens only with the changes in mdx.js file. I'll be so thankful if you can find what is the reason.

**Reply by vahidtakro** [reply]
I just changed it to last mdx.js file you made, it shows my posts now, but only in second language, not first language. also, it shows 404 error when I want to enter the second page on the blog page.
I don't know if the access I gave you is working, I just recreated it.

**Reply by GautierArcin** [reply]
Indeed, there was a bug on the pages. Now everything is fixed and it should be alright. I pushed my modifications.

I did not found any bug on the pages though. It displayed posts in both languages.

**Reply by vahidtakro** [reply]
> Indeed, there was a bug on the pages. Now everything is fixed and it should be alright. I pushed my modifications.
> 
> I did not found any bug on the pages though. It displayed posts in both languages.

Thanks, I can now manage it on both Github and localhost, it's perfect at this moment. 
and I think I found the reason why I can not run it on localhost, I just figured out when I clone my repo to my drive desktop, it works in localhost, so I can see the posts, but when I clone it somewhere else than C drive, where my windows works, it's not showing the posts running in localhost. it's while this wasn't like that before, with Tim repo. does it make sense? or the reason must be something else? at least it works this way.

**Reply by vahidtakro** [reply]
@GautierArcin another bug I may can confirm, is:
if a common .md post (in both languages), use a tag in common, the post in both languages are shown in that [tag] page, if your locale is other than default.
for example, if you put "hello" as tag for a post in both .md and .fr.md files, it will show both files in mywebsite.com/fr/tags/hello

PS: this is only about md, not mdx.
PS2: I just found this is not only about the tags, but also it's the same everywhere, even index page, it's showing posts in English and French if the file is md and your locale is /fr.

I tested this on your repo.

**Reply by GautierArcin** [reply]
I tried to replicate the bug. 

It's working as intented on my repo.

Please see [this branch](https://github.com/GautierArcin/tailwind-nextjs-starter-blog/tree/test/bug_tag_local), or [the live preview](https://tailwind-nextjs-starter-blog-9wb14dqwr-gautierarcin.vercel.app/fr/tags/hello).

Please provide a workable branch with the bug (forked from my repo) if you want me to look into it.

**Reply by vahidtakro** [reply]
@GautierArcin okay I just copied the repo, and added the posts. you can check now:
[live version](https://workingrepo.vercel.app/fr) and [repo](https://github.com/vahidtakro/workingrepo)

**Reply by GautierArcin** [reply]
@vahidtakro  Nice find ! Indeed, there was a bug. 

I fixed it on my repo. I can't change it on yours.

Replace both occurence of 
```javascript
 : getAllFilesRecursively(prefixPaths).filter((path) => path.includes(otherLocale))
```
by
```javascript
 : getAllFilesRecursively(prefixPaths).filter((path) => path.includes(`.${otherLocale}.`))
```

in `lib/mdx.js`

Basically, the filter was not strict enough.

**Reply by vahidtakro** [reply]
> @vahidtakro Nice find ! Indeed, there was a bug.
> 
> I fixed it on my repo. I can't change it on yours.
> 
> Replace both occurence of
> 
> ```js
>  : getAllFilesRecursively(prefixPaths).filter((path) => path.includes(otherLocale))
> ```
> 
> by
> 
> ```js
>  : getAllFilesRecursively(prefixPaths).filter((path) => path.includes(`.${otherLocale}.`))
> ```
> 
> in `lib/mdx.js`
> 
> Basically, the filter was not strict enough.

Thanks! it works very good now!
I just wonder why you added
```
import remarkTocHeadings from './remark-toc-headings'
```
in your code? It's not working for me when I include it. even after installing that package. it works perfectly without it, anyway.
![Captulkre](https://user-images.githubusercontent.com/40500566/128640081-1038e6e3-eb61-4e8e-9824-9aee7e81fb9b.JPG)

**Reply by GautierArcin** [reply]
It's not me that added that line. Please check the commit history.

**Reply by vahidtakro** [reply]
@GautierArcin Thanks Gautier for the sitemap generator, do you think RSS generator and SEO.js must change too?

**Comment by rsipakov** [comment]
@GautierArcin  Regarding newsletter subscription I created a new topic in discussion #177 and provided some ideas for this case on this [branch demo/translate](https://github.com/rsipakov/tailwind-nextjs-starter-blog/tree/demo/translate). 

Also, I created discussion topic #179 for your idea with DeepL API.

**Reply by vahidtakro** [reply]
@rsipakov maybe this one for [Buttondown](https://leerob.io/snippets/buttondown) or this one for [Mailchimp](https://leerob.io/snippets/mailchimp) ?

**Reply by GautierArcin** [reply]
Thank you a lot for both topics you created.

**Reply by rsipakov** [reply]
> @rsipakov maybe this one for [Buttondown](https://leerob.io/snippets/buttondown) or this one for [Mailchimp](https://leerob.io/snippets/mailchimp) ?

Mailchimp, See more here #177

**Comment by vahidtakro** [comment]
Anyone knows what are better commenting system alternatives?
I don't want to ask users to log into github to be able to comment, so they can easily comment just by their emails and names, with a self hosted data / google firebase database. Do you know any packages? I found one but it seems that it's not sending notifications to users when their comments are being replied.

**Reply by timlrx** [reply]
@vahidtakro, I would appreciate it if you could open a new thread for a new topic - keeps things cleaner and makes it easier for people to find similar issues. Anyway, if you had checked out the readme or the config file, you would have seen that disqus is supported out of the box. That's probably the most widely used blog commenting system and supports multiple identity providers beyond Github.

**Reply by vahidtakro** [reply]
> @vahidtakro, I would appreciate it if you could open a new thread for a new topic - keeps things cleaner and makes it easier for people to find similar issues. Anyway, if you had checked out the readme or the config file, you would have seen that disqus is supported out of the box. That's probably the most widely used blog commenting system and supports multiple identity providers beyond Github.

Sure thing! thanks. I will repost it there, 
Yes, I've checked Disqus, it's good , but I think it's not possible to customize? I want a more simple one. anyway I will check Disqus, to make sure, if it's not what I exactly want, I will open a new thread.

**Comment by miryamfv** [comment]
hello, this is exactly what I need, has anyone tried using Tina CMS with this template? I did but got the error: npm `ERR! Failed at the tailwind-nextjs-starter-blog@1.5.6 dev script.` any ideas?
